core/network/src/consensus_gossip.rs

```diff
@@ -40,17 +40,26 @@ struct PeerConsensus<H> {
 	is_authority: bool,
 }
 
+#[derive(Clone, Copy)]
+enum Status {
+	Live,
+	Future,
+}
+
 struct MessageEntry<B: BlockT> {
 	message_hash: B::Hash,
 	topic: B::Hash,
 	message: ConsensusMessage,
 	timestamp: Instant,
+	status: Status,
 }
 
 /// Message validation result.
 pub enum ValidationResult<H> {
 	/// Message is valid with this topic.
 	Valid(H),
+	/// Message is future with this topic.
+	Future(H),
```

Comment:
Why not separate this into `ValidationResult { Valid, Invalid }` and `ExpirationResult { Live, Expired, Future }`? This would make sure that the interface we provide always starts by checking expiration, and only calls `validate` once which presumably might do heavier checks.

Reply:
With the current usage of `validate` in GRANDPA it doesn't make a difference since it always starts by doing the expiration checks, but I think it would make the interface cleaner.

Reply:
(I agree, but could we file a follow-up for that?)